File: packages/java/src/main/java/io/greyfoundry/identifold/storage/mysql/MySQLStorageAdapter.java
package io.greyfoundry.identifold.storage.mysql;

import io.greyfoundry.identifold.Identifold;
import io.greyfoundry.identifold.storage.ReferenceMapping;
import io.greyfoundry.identifold.storage.ReferenceReservation;
import io.greyfoundry.identifold.storage.SequenceAllocationRequest;
import io.greyfoundry.identifold.storage.StorageAdapter;
import java.nio.ByteBuffer;
import java.sql.SQLException;
import java.sql.Types;
import java.util.Objects;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.Executor;
import java.util.regex.Pattern;
import javax.sql.DataSource;

public final class MySQLStorageAdapter implements StorageAdapter {
    private static final Pattern SEQUENTIAL_REFERENCE = Pattern.compile(
        "^([A-Z]{2,8})-(?:([0-9]{4})-)?([0-9]{4,18})-[0-9A-Z*~$=U]$");

    private final DataSource dataSource;
    private final Executor executor;

    public MySQLStorageAdapter(DataSource dataSource, Executor executor) {
        this.dataSource = Objects.requireNonNull(dataSource, "dataSource");
        this.executor = Objects.requireNonNull(executor, "executor");
    }

    @Override
    public CompletionStage<Boolean> reserve(ReferenceReservation request) {
        return CompletableFuture.supplyAsync(() -> {
            try (var connection = dataSource.getConnection();
                 var statement = connection.prepareCall(
                     "CALL identifold_reserve_reference(?, ?, ?)")) {
                statement.setBytes(1, machineIdBytes(request.machineId()));
                statement.setString(2, request.namespace());
                statement.setString(3, request.reference());
                try (var result = statement.executeQuery()) {
                    if (!result.next()) throw failure("allocation_conflict");
                    return result.getBoolean(1);
                }
            } catch (SQLException | IllegalArgumentException exception) {
                throw failure(mapCode(exception));
            }
        }, executor);
    }

    @Override
    public CompletionStage<ReferenceMapping> resolve(String reference, String namespace) {
        return CompletableFuture.supplyAsync(() -> {
            try (var connection = dataSource.getConnection()) {
                var mapping = lookup(
                    connection,
                    "SELECT machine_id, namespace FROM identifold_references "
                        + "WHERE reference = ? AND namespace = ?",
                    namespace,
                    reference,
                    null,
                    null);
                if (mapping != null) return mapping;
                var matcher = SEQUENTIAL_REFERENCE.matcher(reference);
                if (!matcher.matches()) return null;
                return lookup(
                    connection,
                    "SELECT machine_id, namespace FROM identifold_sequence_allocations "
                        + "WHERE namespace = ? AND reference_prefix = ? "
                        + "AND scope = ? AND sequence = ?",
                    namespace,
                    matcher.group(1),
                    matcher.group(2) == null ? "" : matcher.group(2),
                    matcher.group(3));
            } catch (SQLException | IllegalArgumentException exception) {
                throw failure(mapCode(exception));
            }
        }, executor);
    }

    @Override
    public CompletionStage<Long> allocate(SequenceAllocationRequest request) {
        return CompletableFuture.supplyAsync(() -> allocateSynchronously(request), executor);
    }

    private long allocateSynchronously(SequenceAllocationRequest request) {
        for (var attempt = 0; attempt < 5; attempt++) {
            try (var connection = dataSource.getConnection();
                 var statement = connection.prepareCall(
                     "CALL identifold_allocate_sequence(?, ?, ?, ?, ?)")) {
                statement.setBytes(1, machineIdBytes(request.machineId()));
                statement.setString(2, request.namespace());
                statement.setString(3, request.referencePrefix());
                if (request.scope() == null) statement.setNull(4, Types.VARCHAR);
                else statement.setString(4, request.scope());
                statement.setInt(5, request.width());
                try (var result = statement.executeQuery()) {
                    if (!result.next()) throw failure("allocation_conflict");
                    var sequence = result.getLong(1);
                    if (result.wasNull() || sequence < 0) throw failure("allocation_conflict");
                    return sequence;
                }
            } catch (SQLException exception) {
                if (attempt < 4 && isTransient(exception)) {
                    pause(attempt);
                    continue;
                }
                throw failure(mapCode(exception));
            } catch (IllegalArgumentException exception) {
                throw failure("allocation_conflict");
            }
        }
        throw failure("allocation_conflict");
    }

    private static ReferenceMapping lookup(
        java.sql.Connection connection,
        String query,
        String namespace,
        String first,
        String second,
        String third) throws SQLException {
        try (var statement = connection.prepareStatement(query)) {
            if (third == null) {
                statement.setString(1, first);
                statement.setString(2, namespace);
            } else {
                statement.setString(1, namespace);
                statement.setString(2, first);
                statement.setString(3, second);
                statement.setString(4, third);
            }
            try (var result = statement.executeQuery()) {
                if (!result.next()) return null;
                return new ReferenceMapping(
                    bytesMachineId(result.getBytes(1)), result.getString(2));
            }
        }
    }

    private static byte[] machineIdBytes(String value) {
        var uuid = UUID.fromString(value);
        return ByteBuffer.allocate(16)
            .putLong(uuid.getMostSignificantBits())
            .putLong(uuid.getLeastSignificantBits())
            .array();
    }

    private static String bytesMachineId(byte[] value) {
        if (value == null || value.length != 16) throw new IllegalArgumentException();
        var buffer = ByteBuffer.wrap(value);
        return new UUID(buffer.getLong(), buffer.getLong()).toString();
    }

    private static void pause(int attempt) {
        try {
            Thread.sleep(1L << attempt);
        } catch (InterruptedException exception) {
            Thread.currentThread().interrupt();
            throw failure("allocation_conflict");
        }
    }

    private static boolean isTransient(SQLException exception) {
        return exception.getErrorCode() == 1205
            || exception.getErrorCode() == 1213
            || "40001".equals(exception.getSQLState());
    }

    private static String mapCode(Exception exception) {
        if (exception instanceof SQLException sqlException) {
            if ("22003".equals(sqlException.getSQLState())) return "sequence_overflow";
            if ("22023".equals(sqlException.getSQLState())) return "invalid_allocation_policy";
        }
        return "allocation_conflict";
    }

    private static CompletionException failure(String code) {
        return new CompletionException(new Identifold.IdentifoldException(code));
    }
}

File: packages/php/src/Storage/MySQL/MySQLStorageAdapter.php
<?php

declare(strict_types=1);

namespace Greyfoundry\Identifold\Storage\MySQL;

use Greyfoundry\Identifold\IdentifoldException;
use Greyfoundry\Identifold\Storage\ReferenceLookup;
use Greyfoundry\Identifold\Storage\ReferenceMapping;
use Greyfoundry\Identifold\Storage\ReferenceReservation;
use Greyfoundry\Identifold\Storage\ReferenceStore;
use Greyfoundry\Identifold\Storage\SequenceAllocationRequest;
use Greyfoundry\Identifold\Storage\SequenceAllocator;
use PDO;
use PDOException;
use PDOStatement;

final readonly class MySQLStorageAdapter implements ReferenceStore, ReferenceLookup, SequenceAllocator
{
    public function __construct(private PDO $connection)
    {
    }

    public function reserve(ReferenceReservation $request): bool
    {
        try {
            $value = $this->callScalar(
                'CALL identifold_reserve_reference(?, ?, ?)',
                [
                    [$this->machineIdBytes($request->machineId), PDO::PARAM_LOB],
                    [$request->namespace, PDO::PARAM_STR],
                    [$request->reference, PDO::PARAM_STR],
                ],
            );
            if ($value === true || $value === 1 || $value === '1') {
                return true;
            }
            if ($value === false || $value === 0 || $value === '0') {
                return false;
            }
            throw new IdentifoldException('allocation_conflict');
        } catch (PDOException $exception) {
            throw $this->map($exception);
        }
    }

    public function resolve(string $reference, string $namespace): ?ReferenceMapping
    {
        try {
            $row = $this->selectRow(
                'SELECT machine_id, namespace FROM identifold_references '
                . 'WHERE reference = ? AND namespace = ?',
                [[$reference, PDO::PARAM_STR], [$namespace, PDO::PARAM_STR]],
            );
            if ($row !== null) {
                return $this->mapping($row);
            }

            if (preg_match(
                '/^([A-Z]{2,8})-(?:([0-9]{4})-)?([0-9]{4,18})-[0-9A-Z*~$=U]$/D',
                $reference,
                $matches,
            ) !== 1) {
                return null;
            }
            $row = $this->selectRow(
                'SELECT machine_id, namespace FROM identifold_sequence_allocations '
                . 'WHERE namespace = ? AND reference_prefix = ? AND scope = ? AND sequence = ?',
                [
                    [$namespace, PDO::PARAM_STR],
                    [$matches[1], PDO::PARAM_STR],
                    [$matches[2] ?? '', PDO::PARAM_STR],
                    [$matches[3], PDO::PARAM_STR],
                ],
            );
            return $row === null ? null : $this->mapping($row);
        } catch (PDOException $exception) {
            throw $this->map($exception);
        }
    }

    public function allocate(SequenceAllocationRequest $request): int
    {
        $machineId = $this->machineIdBytes($request->machineId);
        for ($attempt = 0; $attempt < 5; ++$attempt) {
            try {
                $value = $this->callScalar(
                    'CALL identifold_allocate_sequence(?, ?, ?, ?, ?)',
                    [
                        [$machineId, PDO::PARAM_LOB],
                        [$request->namespace, PDO::PARAM_STR],
                        [$request->referencePrefix, PDO::PARAM_STR],
                        [$request->scope, $request->scope === null ? PDO::PARAM_NULL : PDO::PARAM_STR],
                        [$request->width, PDO::PARAM_INT],
                    ],
                );
                if ((is_int($value) || (is_string($value) && preg_match('/^\d+$/D', $value) === 1))
                    && (int) $value >= 0) {
                    return (int) $value;
                }
                throw new IdentifoldException('allocation_conflict');
            } catch (PDOException $exception) {
                if ($attempt < 4 && $this->transient($exception)) {
                    usleep(1_000 * (2 ** $attempt));
                    continue;
                }
                throw $this->map($exception);
            }
        }
        throw new IdentifoldException('allocation_conflict');
    }

    /** @param list<array{mixed, int}> $parameters */
    private function callScalar(string $sql, array $parameters): mixed
    {
        $statement = $this->statement($sql, $parameters);
        try {
            $value = $statement->fetchColumn();
            while ($statement->nextRowset()) {
                // Drain every result set produced by CALL before reusing the connection.
            }
            return $value;
        } finally {
            $statement->closeCursor();
        }
    }

    /**
     * @param list<array{mixed, int}> $parameters
     * @return array<string, mixed>|null
     */
    private function selectRow(string $sql, array $parameters): ?array
    {
        $statement = $this->statement($sql, $parameters);
        try {
            $row = $statement->fetch(PDO::FETCH_ASSOC);
            return is_array($row) ? $row : null;
        } finally {
            $statement->closeCursor();
        }
    }

    /** @param list<array{mixed, int}> $parameters */
    private function statement(string $sql, array $parameters): PDOStatement
    {
        $statement = $this->connection->prepare($sql);
        foreach ($parameters as $index => [$value, $type]) {
            $statement->bindValue($index + 1, $value, $type);
        }
        $statement->execute();
        return $statement;
    }

    /** @param array<string, mixed> $row */
    private function mapping(array $row): ReferenceMapping
    {
        if (!is_string($row['machine_id'] ?? null) || !is_string($row['namespace'] ?? null)) {
            throw new IdentifoldException('allocation_conflict');
        }
        return new ReferenceMapping($this->bytesMachineId($row['machine_id']), $row['namespace']);
    }

    private function machineIdBytes(string $value): string
    {
        $hex = str_replace('-', '', $value);
        $bytes = preg_match('/^[0-9a-fA-F]{32}$/D', $hex) === 1 ? hex2bin($hex) : false;
        if ($bytes === false) {
            throw new IdentifoldException('allocation_conflict');
        }
        return $bytes;
    }

    private function bytesMachineId(string $value): string
    {
        if (strlen($value) !== 16) {
            throw new IdentifoldException('allocation_conflict');
        }
        $hex = bin2hex($value);
        return substr($hex, 0, 8) . '-' . substr($hex, 8, 4) . '-'
            . substr($hex, 12, 4) . '-' . substr($hex, 16, 4) . '-' . substr($hex, 20, 12);
    }

    private function transient(PDOException $exception): bool
    {
        $state = $exception->errorInfo[0] ?? $exception->getCode();
        $code = $exception->errorInfo[1] ?? null;
        return $state === '40001' || $code === 1205 || $code === 1213;
    }

    private function map(PDOException $exception): IdentifoldException
    {
        $state = $exception->errorInfo[0] ?? $exception->getCode();
        return new IdentifoldException(match ($state) {
            '22003' => 'sequence_overflow',
            '22023' => 'invalid_allocation_policy',
            default => 'allocation_conflict',
        });
    }
}
